src/NServiceBus.Core/Sagas/SagaMetadata.cs

```diff
@@ -110,5 +119,273 @@ public bool TryGetFinder(string messageType, out SagaFinderDefinition finderDefi
+            {
+                if (result.Any(m => m.MessageType == messageType.FullName))
+                {
+                    continue;
+                }
+                result.Add(new SagaMessage(messageType.FullName, false));
+            }
+
+            return result;
+        }
+
+        static IEnumerable<Type> GetMessagesCorrespondingToFilterOnSaga(Type sagaType, Type filter)
+        {
+            foreach (var interfaceType in sagaType.GetInterfaces())
+            {
+                foreach (var argument in interfaceType.GetGenericArguments())
+                {
+                    var genericType = filter.MakeGenericType(argument);
+                    var isOfFilterType = genericType == interfaceType;
+                    if (!isOfFilterType)
+                    {
+                        continue;
+                    }
+                    yield return argument;
+                }
+            }
+        }
+
+        class SagaMapper : IConfigureHowToFindSagaWithMessage
+        {
+            public List<SagaToMessageMap> Mappings = new List<SagaToMessageMap>();
+
+            void IConfigureHowToFindSagaWithMessage.ConfigureMapping<TSagaEntity, TMessage>(Expression<Func<TSagaEntity, object>> sagaEntityProperty, Expression<Func<TMessage, object>> messageExpression)
+            {
+                var sagaProp = Reflect<TSagaEntity>.GetProperty(sagaEntityProperty, true);
+
+                ValidateMapping<TSagaEntity, TMessage>(messageExpression, sagaProp);
+
+                ThrowIfNotPropertyLambdaExpression(sagaEntityProperty, sagaProp);
+                var compiledMessageExpression = messageExpression.Compile();
+                var messageFunc = new Func<object, object>(o => compiledMessageExpression((TMessage)o));
+
+                Mappings.Add(new SagaToMessageMap
+                {
+                    MessageProp = messageFunc,
+                    SagaPropName = sagaProp.Name,
+                    MessageType = typeof(TMessage)
+                });
+            }
+
+            static void ValidateMapping<TSagaEntity, TMessage>(Expression<Func<TMessage, object>> messageExpression, PropertyInfo sagaProp)
+            {
+                if (sagaProp.Name.ToLower() != "id")
+                {
+                    return;
+                }
+
+                if (messageExpression.Body.NodeType != ExpressionType.Convert)
+                {
+                    return;
+                }
+
+                var memberExpr = ((UnaryExpression)messageExpression.Body).Operand as MemberExpression;
+
+                if (memberExpr == null)
+                {
+                    return;
+                }
+
+                var propertyInfo = memberExpr.Member as PropertyInfo;
+
+                var message = "Message properties mapped to the saga id needs to be of type Guid, please change property {0} on message {1} to a Guid";
+
+                if (propertyInfo != null)
```

Comment:
The following two `if` statements can probably be refactored somehow to remove the repetition